controller/kubernetes_pod_controller.go

```diff
@@ -0,0 +1,289 @@
+	kubeClient clientset.Interface,
+	controllerID string) *KubernetesPodController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	kc := &KubernetesPodController{
+		controllerID: controllerID,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: controllerAgentName}),
+
+		pLister:   kubePodInformer.Lister(),
+		pvLister:  kubePersistentVolumeInformer.Lister(),
+		pvcLister: kubePersistentVolumeClaimInformer.Lister(),
+
+		pStoreSynced:   kubePodInformer.Informer().HasSynced,
+		pvStoreSynced:  kubePersistentVolumeInformer.Informer().HasSynced,
+		pvcStoreSynced: kubePersistentVolumeClaimInformer.Informer().HasSynced,
+
+		queue: workqueue.NewNamedRateLimitingQueue(EnhancedDefaultControllerRateLimiter(), "longhorn-kubernetes-pod"),
+	}
+
+	kubePodInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod := obj.(*v1.Pod)
+			kc.enqueuePodChange(pod)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			curPod := cur.(*v1.Pod)
+			kc.enqueuePodChange(curPod)
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*v1.Pod)
+			kc.enqueuePodChange(pod)
+		},
+	})
+
+	return kc
+}
+
+func (kc *KubernetesPodController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer kc.queue.ShutDown()
+
+	logrus.Infof("Start %v", controllerAgentName)
+	defer logrus.Infof("Shutting down %v", controllerAgentName)
+
+	if !controller.WaitForCacheSync(controllerAgentName, stopCh, kc.pStoreSynced, kc.pvStoreSynced, kc.pvcStoreSynced) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(kc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (kc *KubernetesPodController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesPodController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+	err := kc.syncHandler(key.(string))
+	kc.handleErr(err, key)
+	return true
+}
+
+func (kc *KubernetesPodController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		logrus.Warnf("%v: Error syncing Longhorn kubernetes pod %v: %v", controllerAgentName, key, err)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	logrus.Warnf("%v: Dropping Longhorn kubernetes pod %v out of the queue: %v", controllerAgentName, key, err)
+	kc.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (kc *KubernetesPodController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync %v", controllerAgentName, key)
```

Comment:
Would be good to keep the message consistent with the other controllers, suggest dropping the `controllerAgentName`.

Reply:
same as https://github.com/longhorn/longhorn-manager/pull/640#discussion_r476965100